core/vm/instructions.go

```diff
@@ -547,7 +543,12 @@ func opBlockhash(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) (
 		lower = upper - 256
 	}
 	if num64 >= lower && num64 < upper {
-		num.SetBytes(interpreter.evm.Context.GetHash(num64).Bytes())
+		// if Prague is active, read it from the history contract (EIP 2935).
+		if evm.chainRules.IsPrague {
+			num.SetBytes(getBlockHashFromContract(num64, evm.StateDB, evm.Accesses).Bytes())
+		} else {
+			num.SetBytes(interpreter.evm.Context.GetHash(num64).Bytes())
+		}
```

Comment:
This is a backport of https://github.com/gballet/go-ethereum/pull/391/commits/3be56e9e16c139c3a8edebc2e76b9a332221c1a3 that now is necessary to limit the scope of BLOCKHASH resolving.